### PR TITLE
handle generics in resolve_type()

### PR DESCRIPTION
**Pull request by jjergus**
This changes the public API -- if we care about that, I can instead add this as a new function.

### CODE REVIEW

File: tests/ResolutionTest.hack [hunk 1]
```diff
@@ -187,6 +187,8 @@ final class ResolutionTest extends TestCase {
         'Herp\\Derp',
         'Herp\\Derp',
       ),
+      // generics
+      tuple('<?hh namespace Foo; class Target<T> {}', 'T', 'T'),
```
Comment (jjergus) [inline tests/ResolutionTest.hack]: fails on master

File: src/nodes/Script.hack [hunk 2]
```diff
@@ -86,4 +86,54 @@ final class Script extends ScriptGeneratedBase {
...
+  /**
+   * Find all nodes that have generics and return both the nodes and their
+   * generics.
+   */
+  <<__Memoize>>
+  public function getNodesWithGenerics(): vec<(Node, TypeParameters)> {
+    $root = $this->getDeclarationsx();
+
+    // Sadly these don't have a consistent API so we need a separate Vec\map
+    // for each node type.
```
Comment (fredemmott) [inline src/nodes/Script.hack]: Could be worth adding some - e.g. for functions/methods, IFunctionishDeclaration could be extended - MethodishDeclaration already has manual overrides with a generated base class, FunctionDeclaration would need the same

File: src/resolve_type.hack [hunk 3]
```diff
@@ -12,12 +12,35 @@ namespace Facebook\HHAST;
 use namespace HH\Lib\C;
 use namespace Facebook\HHAST\__Private\Resolution;
 
-function resolve_type(string $type, Script $root, Node $node): string {
+enum ResolvedTypeKind: string {
+  GENERIC_PARAMETER = 'generic_parameter';
+  QUALIFIED_TYPE = 'qualified_type';
+}
+
```
Comment (fredemmott) [inline src/resolve_type.hack]: Please split this to a separate file

Comment (jjergus) [reply src/resolve_type.hack]: Hmm I'm OK either way but in general I don't really like putting everything in separate files -- it feels like taking a class and putting each method in a different file :/

File: src/nodes/Script.hack (hunk 2 above)
Comment (jjergus) [reply src/nodes/Script.hack]: Yeah I tried that first but it turned out to be a lot of changes and this was much simpler. We can always refactor this later.

**Comment by jjergus** [comment]
> This changes the public API -- if we care about that, I can instead add this as a new function.

Looks like [nothing is using this](https://github.com/search?q=org%3Ahhvm+%22resolve_type%22&type=Code) so we should be fine.

**Comment by fredemmott** [comment]
It's also in `__Private`; while we probably want to start using this outside of HHAST, for as long as it's in __Private, breaking BC is fine

**Comment by jjergus** [comment]
It's not -- only `resolve_name` (the implementation shared by `resolve_type` and `resolve_function` is __Private)

**Comment by fredemmott** [comment]
oh whoops; thanks for catching that and checking for callers :)

**Pull request merged**